specs/saltpack_signcryption_v2.md

```diff
@@ -0,0 +1,278 @@
+attacks.
+
+As in the encryption mode, each encrypted copy of the message encryption key is
+given a recipient identifier. Unlike the encryption mode, signcryption doesn't
+use the literal public key bytes as the identifier (in the case of an
+asymmetric recipient), but instead an opaque identifier derived from the
+recipient's key and the ephemeral sender key. This preserves recipient
+anonymity by default. Applications lose the ability to make suggestions like
+"your phone can't read this message, but your laptop can" based on the
+recipients list, but the intention is that distributing long-term shared
+symmetric keys to all of a user's devices can avoid that situation entirely.
+The recipient identifiers for symmetric recipients are entirely up to the
+application.
+
+The signcryption scheme allows an anonymous sender by specifying a signing key
+that is all zeroes. In this case the signature is filled with zeroes also, and
+the verification step is skipped. This would be similar to using the encryption
+mode with an anonymous sender and a shared asymmetric recipient key, though the
+encryption mode wasn't designed with shared recipient keys in mind. (In
+particular it would be very wrong to share recipient keys in the encryption
+mode, if the sender was authenticated).
+
+## Implementation
+
+An signcrypted message is a series of concatenated MessagePack objects. The
+first is a header packet, followed by one or more payload packets, the last of
+which is indicated with a final packet flag.
+
+### Header Packet
+The header packet is a MessagePack list with these contents:
+
+```
+[
+    format name,
+    version,
+    mode,
+    ephemeral public key,
+    sender secretbox,
+    recipients list,
+]
+```
+
+- The **format name** is the string "saltpack".
+- The **version** is a list of the major and minor versions, currently
+  `[2, 0]`. Note that saltpack version 1 did not include a signcryption mode.
+- The **mode** is the number 3, for encryption. (1 and 2 are attached and
+  detached signing, and 3 is signcryption.)
+- The **ephemeral public key** is a NaCl public encryption key, 32 bytes. The
+  ephemeral keypair is generated at random by the sender and only used for one
+  message.
+- The **sender secretbox** is a
+  [`crypto_secretbox`](http://nacl.cr.yp.to/secretbox.html) containing the
+  sender's long-term public _signing_ key, encrypted with the **payload key**
+  from below.
+- The **recipients list** contains a recipient pair for each recipient key,
+  including an encrypted copy of the **payload key**. See below.
+
+A recipient pair is a two-element list:
+
+```
+[
+    recipient identifier,
+    payload key box,
+]
+```
+
+- The **recipient identifier** depends on the type of recipient key. For an
+  asymmetric recipient key, the identifier is derived from shared DH output.
+  For a symmetric key, the identifier is left to the application.
+- The **payload key box** is a [`crypto_box`](http://nacl.cr.yp.to/box.html)
```

Comment:
Going to land as is, but let me know what you think about these details, and I can make a followup PR.